lib/unison-sqlite/src/Unison/Sqlite/Transaction.hs

```diff
@@ -67,13 +71,33 @@ newtype Transaction a
 
 -- | Run a transaction on the given connection.
 runTransaction :: MonadIO m => Connection -> Transaction a -> m a
-runTransaction conn@(Connection _ _ conn0) (Transaction f) =
-  -- TODO some sensible retry logic
-  liftIO do
-    Sqlite.execute_ conn0 "BEGIN"
-    result <- f conn
-    Sqlite.execute_ conn0 "COMMIT"
+runTransaction conn (Transaction f) = liftIO do
+  uninterruptibleMask \restore -> do
+    Connection.execute_ conn "BEGIN"
+    result <-
+      try (restore (f conn)) >>= \case
+        Left exception -> do
+          rollback
+          case exception of
+            SqliteBusyException ->
+              (`fix` 100_000) \again microseconds -> do
+                restore (threadDelay microseconds)
+                try (Connection.execute_ conn "BEGIN IMMEDIATE") >>= \case
+                  Left SqliteBusyException -> again (microseconds * 2)
+                  Left exception -> throwIO exception
+                  Right () -> restore (f conn) `onException` ignoringExceptions rollback
+            _ -> throwIO exception
+        Right result -> pure result
+    Connection.execute_ conn "COMMIT"
     pure result
+  where
+    rollback :: IO ()
+    rollback =
+      Connection.execute_ conn "ROLLBACK"
+
+    ignoringExceptions :: IO () -> IO ()
+    ignoringExceptions action =
+      action `catchAny` \_ -> pure ()
```

Comment:
This seems to just silently swallow and ignore any exception here. I think it's sufficiently unlikely that this will throw an exception and that we should allow it to propagate so we at least know that it's happening rather than failing to rollback and silencing any possible issues.

Reply:
I disagree. I think if the user-provided action throws an exception, and we have a transaction open, then we should attempt a rollback, and regardless of if it succeeds or not, propagate the original exception.

Reply:
Ah, I didn't see that this still propagates the original exception.

I think that makes sense to propagate the original exception, but I think we should likely still print out or log the failure to roll-back.